service/history/queues/slice.go

```diff
@@ -0,0 +1,370 @@
+	currentIterIdx := 0
+	incomingIterIdx := 0
+	for currentIterIdx < len(s.iterators) && incomingIterIdx < len(incomingSlice.iterators) {
+		currentIter := s.iterators[currentIterIdx]
+		incomingIter := incomingSlice.iterators[incomingIterIdx]
+
+		if currentIter.Range().InclusiveMin.CompareTo(incomingIter.Range().InclusiveMin) < 0 {
+			mergedIterators = s.appendIterator(mergedIterators, currentIter)
+			currentIterIdx++
+		} else {
+			mergedIterators = s.appendIterator(mergedIterators, incomingIter)
+			incomingIterIdx++
+		}
+	}
+
+	for _, iterator := range s.iterators[currentIterIdx:] {
+		mergedIterators = s.appendIterator(mergedIterators, iterator)
+	}
+	for _, iterator := range incomingSlice.iterators[incomingIterIdx:] {
+		mergedIterators = s.appendIterator(mergedIterators, iterator)
+	}
+
+	return mergedIterators
+}
+
+func (s *SliceImpl) appendIterator(
+	iterators []Iterator,
+	iterator Iterator,
+) []Iterator {
+	if len(iterators) == 0 {
+		return []Iterator{iterator}
+	}
+
+	size := len(iterators)
+	if iterators[size-1].CanMerge(iterator) {
+		iterators[size-1] = iterators[size-1].Merge(iterator)
+		return iterators
+	}
+
+	return append(iterators, iterator)
+}
+
+func (s *SliceImpl) ShrinkRange() {
+	s.validateNotDestroyed()
+
+	minTaskKey := tasks.MaximumKey
+	for key := range s.outstandingExecutables {
+		if s.outstandingExecutables[key].State() == ctasks.TaskStateAcked {
+			delete(s.outstandingExecutables, key)
+			continue
+		}
+
+		minTaskKey = tasks.MinKey(minTaskKey, key)
+	}
+
+	// still has pending task in memory
+	if len(s.outstandingExecutables) != 0 {
+		s.scope.Range.InclusiveMin = minTaskKey
+		return
+	}
+
+	// no more pending task in memory, but has more tasks to read from persistence
+	if len(s.iterators) != 0 {
+		s.scope.Range.InclusiveMin = s.iterators[0].Range().InclusiveMin
+		return
+	}
+
+	// no pending task in memory and persistence
+	s.scope.Range.InclusiveMin = s.scope.Range.ExclusiveMax
+}
+
+func (s *SliceImpl) SelectTasks(batchSize int) ([]Executable, error) {
+	s.validateNotDestroyed()
+
+	if len(s.iterators) == 0 {
+		return []Executable{}, nil
+	}
+
+	executables := make([]Executable, 0, batchSize)
+	for len(executables) < batchSize && len(s.iterators) != 0 {
+		if s.iterators[0].HasNext() {
+			task, err := s.iterators[0].Next()
+			if err != nil {
+				s.iterators[0] = s.iterators[0].Remaining()
+				return nil, err
+			}
+
+			taskKey := task.GetKey()
+			if !s.scope.Range.ContainsKey(taskKey) {
+				panic(fmt.Sprintf("Queue slice get task from iterator doesn't belong to its range, range: %v, task key %v",
+					s.scope.Range, taskKey))
+			}
+
+			if !s.scope.Predicate.Test(task) {
+				continue
+			}
+
+			executable := s.executableInitializer(task)
+			s.outstandingExecutables[taskKey] = executable
+			executables = append(executables, executable)
+		} else {
+			s.iterators = s.iterators[1:]
+		}
+	}
+
+	return executables, nil
+}
+
+func (s *SliceImpl) destroy() {
+	s.destroyed = true
```

Comment:
do you need to release resource? (like the map and slice in it)

Reply:
I was thinking caller will no longer reference the slice after it's destroyed. But yeah, it's better to do it explicitly here so we won't rely on caller behavior.